### PR TITLE
Chain id and version in transaction structure

### PR DESCRIPTION
**Pull request by miiu96**
Added `chain ID`  and `version` in transaction structure. Now all transaction that are generated (from wallet, erdpy, tx-gen etc must contains correct `chain ID` and `version` in order to be accepted by the observers). Also the transaction must be signed with the `chain ID` and `version`.

### CODE REVIEW

File: integrationTests/frontend/wallet/txInterception_test.go [hunk 1]
```diff
@@ -39,11 +40,12 @@ func TestInterceptedTxWhithDataField(t *testing.T) {
...
 		value,
 		"erd12dnfhej64s6c56ka369gkyj3hwv5ms0y5rxgsk2k7hkd2vuk7rvqxkalsa",
-		"erd1zq6l0g5yg6ehz7reatt3uhyfy89udx8q3p72fy5fhmfdvr883xgqufyvte",
-		"8bdaebfc67b5b7e8a325004b3934dc6f1f0329b3d29baa1763c663925384bd93f1de3a8ac297db918a233c13c4014431a8ed7455c03e2a706828a9a400b6b107",
+		"erd1xehvyp68ufu7gd543nrjs4m5fx2d090qkeqh5k7y3yt42en0ealszcr5n8",
+		"9fcccee9545345be4aa36883b5938581f9d1ff7eb54190b1507364d43aaa8236fe655ef75d77c8973779c321ff454cb35c6689947c81ebbc2208327b694a0708",
 		10,
 		100000,
-		[]byte("data@~`!@#$^&*()_=[]{};'<>?,./|<>><!!!!!"),
+		[]byte("erd-test-data"),
```
Comment (iulianpascalau) [inline integrationTests/frontend/wallet/txInterception_test.go:47]: you discarded the special characters. Does they still work?

File: process/transaction/interceptedTransaction.go [hunk 2]
```diff
@@ -216,6 +222,9 @@ func (inTx *InterceptedTransaction) processFields(txBuff []byte) error {
 
 // integrity checks for not nil fields and negative value
 func (inTx *InterceptedTransaction) integrity(tx *transaction.Transaction) error {
+	if tx.ChainID == nil || !bytes.Equal(tx.ChainID, inTx.chainID) {
```
Comment (iulianpascalau) [inline process/transaction/interceptedTransaction.go:225]: do this directly: 
```GO
if !bytes.Equal(tx.ChainID, inTx.chainID) {
```

File: integrationTests/frontend/wallet/txInterception_test.go (hunk 1 above)
Comment (bogdan-rosianu) [reply integrationTests/frontend/wallet/txInterception_test.go:47]: they work. I just generated something different for Mihai to test and forgot to add back that data field. Will be updated

Comment (miiu96) [reply integrationTests/frontend/wallet/txInterception_test.go:47]: added again special characters

File: process/transaction/interceptedTransaction.go (hunk 2 above)
Comment (miiu96) [reply process/transaction/interceptedTransaction.go:225]: done.

File: api/mock/facade.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ type Facade struct {
 	GenerateTransactionHandler func(sender string, receiver string, value *big.Int, code string) (*transaction.Transaction, error)
 	GetTransactionHandler      func(hash string) (*transaction.ApiTransactionResult, error)
 	CreateTransactionHandler   func(nonce uint64, value string, receiverHex string, senderHex string, gasPrice uint64,
-		gasLimit uint64, data string, signatureHex string, chainID string) (*transaction.Transaction, []byte, error)
+		gasLimit uint64, data string, signatureHex string, chainID string, version uint32) (*transaction.Transaction, []byte, error)
```
Comment (iulianpascalau) [inline api/mock/facade.go:29]: do not forget to expose the minimum transaction version on the network config route (in order to be used by the frontend). Also, create PR to update the proxy/txgen repos. Frontend should be upgraded too.

Comment (sasurobert) [reply api/mock/facade.go:29]: Add unit tests all over the PR where the new data field are verified.

Comment (miiu96) [reply api/mock/facade.go:29]: Already exposed minimum transaction version in api route `/network/config`. And  PR for proxy and tx-gen will be open soon (work in progress).

File: process/factory/interceptorscontainer/shardInterceptorsContainerFactory.go [hunk 4]
```diff
@@ -76,6 +76,12 @@ func NewShardInterceptorsContainerFactory(
 	if check.IfNil(args.EpochStartTrigger) {
 		return nil, process.ErrNilEpochStartTrigger
 	}
+	if len(args.ChainID) == 0 {
+		return nil, process.ErrInvalidChainID
```
Comment (iulianpascalau) [inline process/factory/interceptorscontainer/shardInterceptorsContainerFactory.go:80]: Please treat those 2 errors in multidata and single data interceptor and blacklist the fromconnected peer and message peer originator

File: api/mock/facade.go (hunk 3 above)
Comment (miiu96) [reply api/mock/facade.go:29]: more unit tests was added

File: process/factory/interceptorscontainer/shardInterceptorsContainerFactory.go (hunk 4 above)
Comment (iulianpascalau) [reply process/factory/interceptorscontainer/shardInterceptorsContainerFactory.go:80]: In multidatainterceptor.go, L186 and in singledatainterceptor.go L95 please add:
```GO
		isWrongVersion := err == process.ErrInvalidTransactionVersion || err == process.ErrInvalidChainID
		if isWrongVersion{
			//this situation is so severe that we need to black list de peers
			reason := "wrong version of received intercepted data, topic " + mdi.topic + ", error " + err.Error()
			mdi.antifloodHandler.BlacklistPeer(originator, reason, core.InvalidMessageBlacklistDuration)
			mdi.antifloodHandler.BlacklistPeer(fromConnectedPeer, reason, core.InvalidMessageBlacklistDuration)
		} 
```

please add tests for these condition. Thanks

Comment (miiu96) [reply process/factory/interceptorscontainer/shardInterceptorsContainerFactory.go:80]: done.

File: process/interceptors/multiDataInterceptor_test.go [hunk 5]
```diff
@@ -464,6 +464,75 @@ func TestMultiDataInterceptor_ProcessReceivedMessageWhitelistedShouldRetNil(t *t
...
+	processReceivedMessageMultiDataInvalidVersion(t, process.ErrInvalidTransactionVersion)
+}
+
+func TestMultiDataInterceptor_InvalidTxChainIDShouldBackList(t *testing.T) {
+	t.Parallel()
+
+	processReceivedMessageMultiDataInvalidVersion(t, process.ErrInvalidChainID)
+}
+
+func processReceivedMessageMultiDataInvalidVersion(t *testing.T, expectedErr error) {
```
Comment (iulianpascalau) [inline process/interceptors/multiDataInterceptor_test.go:479]: :+1:

File: process/interceptors/multiDataInterceptor_test.go [hunk 6]
```diff
@@ -464,6 +464,75 @@ func TestMultiDataInterceptor_ProcessReceivedMessageWhitelistedShouldRetNil(t *t
...
+				return interceptedData, nil
+			},
+		},
+		createMockInterceptorStub(&checkCalledNum, &processCalledNum),
+		throttler,
+		&mock.P2PAntifloodHandlerStub{
+			BlacklistPeerCalled: func(peer core.PeerID, reason string, duration time.Duration) {
+				switch string(peer) {
+				case string(originator):
+					isOriginatorBlackListed = true
```
Comment (iulianpascalau) [inline process/interceptors/multiDataInterceptor_test.go:511]: :+1:

**Pull request merged**